File: distros/z/zxid/zxidhlo.php
<?
# zxid/zxidhlo.php  -  Hello World SAML SP role in PHP using zxid extension
#
# Copyright (c) 2010 Sampo Kellomaki (sampo@iki.fi), All Rights Reserved.
# Copyright (c) 2007-2009 Symlabs (symlabs@symlabs.com), All Rights Reserved.
# Author: Sampo Kellomaki (sampo@iki.fi)
# This is confidential unpublished proprietary source code of the author.
# NO WARRANTY, not even implied warranties. Contains trade secrets.
# Distribution prohibited unless authorized in writing.
# Licensed under Apache License 2.0, see file COPYING.
# $Id: zxidhlo.php,v 1.11 2009-11-29 12:23:06 sampo Exp $
# 16.1.2007,  created --Sampo
# 25.5.2008,  fixed to work against 0.27, fixed port number to 5443 --Sampo
# 14.11.2009, Added zxid_az() example --Sampo
# 27.11.2009, Added zxid_call() examples --Sampo

dl("php_zxid.so");  # These three lines can go to initialization: they only need to run once
# CONFIG: You must have created /var/zxid directory hierarchy. See `make dir'
# CONFIG: You must edit the URL to match your domain name and port
#$conf = "URL=https://sp1.zxidsp.org:5443/zxidhlo.php&PATH=/var/zxid/";
$conf = "PATH=/var/zxid/&URL=http://sp.tas3.pt:8082/zxidhlo.php";
#error_log("zxidphp: conf($conf)");
$cf = zxid_new_conf_to_cf($conf);
?>
<?
# For every page that is accessed. Debug: QUERY_STRING=o=E REQUEST_METHOD=GET ./zxidhlo.php
#print_r(phpinfo());
#print_r($_SERVER);
$qs = $_SERVER['REQUEST_METHOD'] == 'GET'
      ? $_SERVER['QUERY_STRING']
      : file_get_contents('php://input');
error_log("zxidphp: qs($qs)");
$res = zxid_simple_cf($cf, -1, $qs, null, 0x1814);
error_log("zxidphp: res($res) conf($conf)");

switch (substr($res, 0, 1)) {
case 'L': header($res); exit;  # Redirect (Location header)
case '<': header('Content-type: text/xml'); echo $res; exit;  # Metadata or SOAP
case 'n': exit;   # Already handled
case 'e':
?>
<title>ZXID PHP Demo Please Login Using IdP</title>
<body bgcolor="#330033" text="#ffaaff" link="#ffddff"
 vlink="#aa44aa" alink="#ffffff"><font face=sans>
<h1>ZXID PHP Demo Please Login Using IdP</h1>
<?=zxid_idp_select_cf($cf, null, 0x1900)?>
<hr>zxidhlo.php, <a href="http://zxid.org/">zxid.org</a>
<?
exit;
case 'd': break;  # Logged in case -- continue after switch
default:  die("Unknown zxid_simple() res($res)");
}

# Parse the LDIF in $res into a hash of attributes $attr

foreach (split("\n", $res) as $line) {
    $a = split(": ", $line);
    $attr[$a[0]] = $a[1];
}
?>
<title>ZXID PHP Demo Protected content, logged in</title>
<body bgcolor="#330033" text="#ffaaff" link="#ffddff"
 vlink="#aa44aa" alink="#ffffff"><font face=sans>
<?

# Optional: Perform additional authorization step
# (n.b. zxid_simple() can be configured to make az automatically)

$ses = zxid_fetch_ses($cf, $attr['sesid']);

if (zxid_az_cf_ses($cf, "Action=Show", $ses)) {
    echo "Permit.\n";
} else {
    echo "<b>Deny.</b> Normally page would not be shown, but we show session attributes for debugging purposes.\n";
}
?>
<h1>ZXID PHP Demo Protected content, logged in as <?=$attr['cn']?>, session(<?=$attr['sesid']?>)</h1>
<?=zxid_fed_mgmt_cf($cf, null, -1, $attr['sesid'], 0x1900)?>

<p>Output from idhrxml web service call:<br><textarea cols=80 rows=20>
<?
$ret = zxid_call($cf, $ses, "urn:id-sis-idhrxml:2007-06:dst-2.1", null, null, null,
		 "<idhrxml:Query>" .
		   "<idhrxml:QueryItem>" .
		     "<idhrxml:Select></idhrxml:Select>" .
		   "</idhrxml:QueryItem>" .
		 "</idhrxml:Query>");
echo $ret;
?>
</textarea>
<p>Output from foobar web service call:<br><textarea cols=80 rows=20>
<?
$ret = zxid_call($cf, $ses, "urn:x-foobar", null, null, null, "<foobar>Do it!</foobar>");
echo $ret;
?>
</textarea>
<p>Output from foobar-php web service call:<br><textarea cols=80 rows=20>
<?
$ret = zxid_call($cf, $ses, "urn:x-foobar-php", null, null, null, "<foobar>Do it!</foobar>");
echo $ret;
?>
</textarea>
<hr>zxidhlo.php, <a href="http://zxid.org/">zxid.org</a>

File: distros/z/zxid/zxidhlo.pl
#!/usr/bin/perl
# Copyright (c) 2007-2009 Symlabs (symlabs@symlabs.com), All Rights Reserved.
# Author: Sampo Kellomaki (sampo@iki.fi)
# This is confidential unpublished proprietary source code of the author.
# NO WARRANTY, not even implied warranties. Contains trade secrets.
# Distribution prohibited unless authorized in writing.
# Licensed under Apache License 2.0, see file COPYING.
# $Id: zxidhlo.pl,v 1.7 2009-11-14 22:44:43 sampo Exp $
# 16.1.2007, created --Sampo
# 14.11.2009, Added zxid_az() example --Sampo

use Net::SAML;
use Data::Dumper;

$| = 1;
undef $/;

open STDERR, ">>tmp/zxid.stderr";   # Helps CGI debugging where web server eats the stderr

#$url = "https://sp1.zxidsp.org:8443/zxidhlo.pl";  # Edit to match your situation
$url = "http://sp.tas3.pt:8082/zxidhlo.pl";  # Edit to match your situation
$conf = "PATH=/var/zxid/&URL=$url";
$cf = Net::SAML::new_conf_to_cf($conf);
#warn "cf($cf):".Dumper($cf);
$qs = $ENV{'QUERY_STRING'};
$qs = <STDIN> if $qs =~ /o=P/;
$res = Net::SAML::simple_cf($cf, -1, $qs, undef, 0x1828);
$op = substr($res, 0, 1);
if ($op eq 'L' || $op eq 'C') { warn "res($res) len=".length($res); print $res; exit; } # LOCATION (Redir) or CONTENT
if ($op eq 'n') { exit; } # already handled
if ($op eq 'e') { my_render_login_screen(); exit; }
if ($op ne 'd') { die "Unknown Net::SAML::simple() res($res)"; }

# *** add code to parse the LDIF in $res into a hash of attributes

($sid) = $res =~ /^sesid: (.*)$/m;

if (Net::SAML::az_cf($cf, "Action=Show", $sid)) {
    $az = "Permit.\n";
} else {
    $az = "<b>Deny.</b> Normally page would not be shown, but we show session attributes for debugging purposes.\n";
}

print <<HTML
CONTENT-TYPE: text/html

<title>ZXID perl HLO SP Mgmt</title>
<link rel="shortcut icon" href="/favicon.ico" type="image/x-icon" />
<body bgcolor="#330033" text="#ffaaff" link="#ffddff" vlink="#aa44aa" alink="#ffffff"><font face=sans>
$az
<h1>ZXID SP Perl HLO Management (user logged in, session active)</h1>
sesid: $sid
HTML
    ;
print Net::SAML::fed_mgmt_cf($cf, undef, -1, $sid, 0x1900);
exit;

###
### Render the login screen
###

sub my_render_login_screen {
    print <<HTML;
CONTENT-TYPE: text/html

<title>ZXID SP PERL HLO SSO</title>
<link rel="shortcut icon" href="/favicon.ico" type="image/x-icon" />
<body bgcolor="#330033" text="#ffaaff" link="#ffddff" vlink="#aa44aa" alink="#ffffff"><font
face=sans><h1>ZXID SP Perl HLO Federated SSO (user NOT logged in, no session.)</h1>
<form method=get action="zxidhlo.pl">

<h3>Login Using New IdP</h3>

<i>A new IdP is one whose metadata we do not have yet. We need to know
the Entity ID in order to fetch the metadata using the well known
location method. You will need to ask the adminstrator of the IdP to
tell you what the EntityID is.</i>

<p>IdP URL <input name=e size=60>
<input type=submit name=l1 value=" Login (A2) ">
<input type=submit name=l2 value=" Login (P2) ">
HTML
;
    print Net::SAML::idp_list_cf($cf, undef, 0x1c00);   # Get the IdP selection form
    print <<HTML;
<h3>CoT configuration parameters your IdP may need to know</h3>

Entity ID of this SP: <a href="$url?o=B">$url?o=B</a> (Click on the link to fetch SP metadata.)

<h3>Technical options</h3>
<input type=checkbox name=fc value=1 checked> Create federation,
   NID Format: <select name=fn>
                 <option value=prstnt>Persistent
                 <option value=trnsnt>Transient
                 <option value="">(none)
               </select><br>

<input type=hidden name=fq value="">
<input type=hidden name=fy value="">
<input type=hidden name=fa value="">
<input type=hidden name=fm value="">
<input type=hidden name=fp value=0>
<input type=hidden name=ff value=0>

</form><hr><a href="http://zxid.org/">zxid.org</a>
HTML
    ;
}

__END__
